### PR TITLE
Test FSDP with submodule non-reentrant checkpointing

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89781

With combining FSDP with reentrant checkpointing, the post backward
hook might run twice, and then hit [this
error](https://github.com/pytorch/pytorch/blob/e20ec44544c17d6d3d411f88b870e05043bda731/torch/distributed/fsdp/_runtime_utils.py#L487).
This is because reentrant backward uses nested autograd GraphTasks.
The inner GraphTask is not aware of the outer one and therefore
will flush pending `AccumulateGrad` invocations on exit, which in
turn triggers the post backward hooks registered by FSDP. Later,
the outer GraphTask will trigger that again, leading to the above
error.

PR #89791 relaxes the FSDP training state check, but we still run
into grad value check failures occasionally. Therefore, this PR only
lands the test for non-reentrant test, and we can enable the
reentrant test when the accuracy issues are addressed.

### CODE REVIEW

File: test/distributed/fsdp/test_fsdp_checkpoint.py
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_checkpoint.py:331]: why don't we just pass `use_reentrant=False` in L325?

File: test/distributed/fsdp/test_fsdp_checkpoint.py
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_checkpoint.py:282]: nit: Should we add a non-linearity (e.g. ReLU) to this model just to be safe for the gradient correctness check between 'with AC' and 'without AC'?

I do not think it matters for this case, but I just generally avoid the case where the model is equivalent to one affine transformation since I have had silent issues before.

File: test/distributed/fsdp/test_fsdp_checkpoint.py
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_checkpoint.py:322]: I am guessing that trying `use_orig_params=True` can be done in a follow-up. This PR only targets `use_orig_params=False` and `NO_SHARD`?

File: test/distributed/fsdp/test_fsdp_checkpoint.py
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_checkpoint.py:335]: nit:
```
fsdp_kwargs = {
    "device_id": torch.cuda.current_device(),
    "sharding_strategy": ShardingStrategy.NO_SHARD,
} 
```
and pass `**fsdp_kwargs` to each call to `FSDP()`

**Review by awgu — COMMENTED** [review]
Thanks for adding these tests!

At a meta level, I am wondering how we should approach testing more interleavings in a systematic and complete way.

Also, I am not sure if you want to wait for the assert relaxation PR to land and then update `test_checkpoint_submodule_reentrant()` or not.

File: test/distributed/fsdp/test_fsdp_checkpoint.py
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_checkpoint.py:314]: nit: instead of names `m1` and `m2`, checkpoint_1 and checkpoint_2 might make things clearer for readers.

File: test/distributed/fsdp/test_fsdp_checkpoint.py
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_checkpoint.py:384]: nit: `device='cuda'`

File: test/distributed/fsdp/test_fsdp_checkpoint.py
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_checkpoint.py:360]: nit: `device='cuda'`

**Review by rohan-varma — APPROVED** [review]
LGTM, stamping to unblock. Will file a follow up issue to debug why this FSDP + AC structure does not work.

File: test/distributed/fsdp/test_fsdp_checkpoint.py
Comment (mrshenli) [reply test/distributed/fsdp/test_fsdp_checkpoint.py:314]: sure, let me update that

File: test/distributed/fsdp/test_fsdp_checkpoint.py
Comment (mrshenli) [reply test/distributed/fsdp/test_fsdp_checkpoint.py:322]: yep, only targeting a specific case at this moment

**Comment by mrshenli** [comment]
> Also, I am not sure if you want to wait for the assert relaxation PR to land and then update test_checkpoint_submodule_reentrant() or not.

Updated the PR summary to include that.

**Pull request closed without merging**
